### ISSUE TITLE
Cannot change color of agents 

### ISSUE THREAD

**Issue by AlonBenAriVA** [scottpdo/flocc#19]
**Describe the bug**

Working off the flocking example seek to change color of agents when they are close and reset when the distance. I cannot get to change the color of the agents, and I am getting an error on the console saying the color is None.
Please advise, what I am doing wrong. Attaching code and screen shot. The agents are very close but color does not change. (Ubuntu 14.04, Mozilla Firefox 66.0.3)
![Screenshot from 2019-11-14 20:59:23](https://user-images.githubusercontent.com/37850639/68918204-1886ee80-0722-11ea-9693-d03ecf5900e2.png)

`const { Agent, Environment, GridEnvironment, ASCIIRenderer, CanvasRenderer, utils, Vector } = flocc;

/* ------- PARAMETERS --------- */

const flockSize = 10;
const width = 300;
const height = 300;

/* ---------------------------- */

/* ------- SET UP ENVIRONMENT, RENDERER --------- */

const environment = new Environment();
const renderer = new CanvasRenderer(environment, { width, height });
const container = document.getElementById('container');
console.log(container)
renderer.mount(container);

function setup() {
  for (let i = 0; i < flockSize; i++) {
    const agent = new Agent();
    agent.set('x', 0.25 * Math.random() * width);
    agent.set('y', 0.25 * Math.random() * height);
    agent.set('shape', 'arrow');
    agent.set('color', '#0F0');
    agent.set('size', 3);
    const angle = 2 * Math.random() * Math.PI;
    agent.set('vx', Math.cos(angle));
    agent.set('vy', Math.sin(angle));

    agent.addRule(tick);

    environment.addAgent(agent);
  }

}

function tick(agent) {
  var x = agent.get('x');
  var y = agent.get('y');
  x +=  agent.get('vx');
  y += agent.get('vy');

  while (x < 0) x += width;
  while (x >= width) x -= width;
  while (y < 0) y += height;
  while (y >= height) y -= height;

  agent.set('x', x);
  agent.set('y', y);

  // update color if the agents are close

  environment.getAgents().forEach(function(x){
    console.log(x.get('color'))
  })
  
  environment.getAgents().filter(function(other){
    var d = utils.distance(agent,other);
    
    if ( d<=60 & d >0 ){
            other.set('color','red');
            agent.set('color','green');
      
    } else {
            other.set('color','blue');
            agent.set('color','blue');
    }
      
      
  });

}

function render() {
  environment.tick();
  requestAnimationFrame(render);
}

setup();
render();
`

**To Reproduce**

**Expected behavior**

**Screenshots**

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**

**Additional context**

**Comment by scottpdo** [comment]
@AlonBenAriVA It looks like there's a couple of issues inside your code that loops over all the other agents. Since it runs across every agent in the environment, for every agent in the environment, the order matters and it will override previous colors that have been set. One way to fix it would be to reset an agent's color before looping over all the others, and then only changing it if it finds another agent within 60 units, like this:

```
agent.set('color', 'blue');
environment.getAgents().filter(function(other){
  var d = utils.distance(agent,other);

  if ( d<=60 && d >0 ) {
    other.set('color', 'red');
    agent.set('color', 'green');
  }
});
```

One potential issue is that it will be difficult to consistently get red _and_ green agents with this code. For example, if A and B are 10 units apart, then looping over A, B will set red and A will set green. Then, it will loop over B and reverse those colors — A red and B green. So it will end up undoing the changes of previous agents.

Unfortunately, I'm not sure why you're seeing the `Expected color but found 'None'` error, but hopefully the above fix can start to address your issue...

**Comment by AlonBenAriVA** [comment]
Firstly, thank you.
For my understanding: flocc iterates over the agent set at every step?
Is there a way to change the order of agents being stepped into ? is it random?, it is possible to set it ?

**Comment by scottpdo** [comment]
For every `tick` of the environment, right now it runs over all the agents in the order they were added in your `setup` function. Then it does the same thing for each agent when you call `environment.getAgents().forEach`. With agents A, B, and C this would look like:
```js
// loop over A
// while looping over A, loop over A, B, and C
// loop over B
// while looping over B, loop over A, B, and C
// loop over C
// while looping over C, loop over A, B, and C
```

It would be possible to randomize the inner loop by using the `utils.shuffle` function, which randomizes the order of an array, which would look like: `utils.shuffle(environment.getAgents()).forEach`. But right now there isn't a way to randomize the order of the outer loop (that gets called with each `environment.tick`). Does that all make sense?

I'm thinking this would be a great feature for the next release, though, maybe by adding configuration options to the `tick` function.

**Comment by scottpdo** [comment]
Hi @AlonBenAriVA — since chatting last I've released [v0.3.1](https://github.com/scottpdo/flocc/releases/tag/v0.3.1), which lets you randomize the order in `environment.tick` by calling it like this:
```js
environment.tick({
  randomizeOrder: true
});
```

Let me know if this helps!

**Comment by scottpdo** [comment]
Closing this one since the 0.3.1 release addresses it. @AlonBenAriVA please let me know if it's still an issue for you.
